core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -0,0 +1,2103 @@
+
+            if (partition.getRemainingAttempts() == 0 || (errorCode != null && errorCode.getType() == USER_ERROR)) {
+                stage.fail(failure);
+                // stage failed, don't reschedule
+                return ImmutableList.of();
+            }
+
+            if (partition.getOpenTaskDescriptor().isPresent()) {
+                // don't reschedule speculative tasks
+                return ImmutableList.of();
+            }
+
+            // TODO: split into smaller partitions here if necessary (for example if a task for a given partition failed with out of memory)
+
+            // reschedule a task
+            return ImmutableList.of(new ScheduledTask(stage.getStageId(), partitionId, schedulingPriority));
+        }
+
+        public Optional<NodeRequirements> getNodeRequirements(int partitionId)
+        {
+            StagePartition partition = getStagePartition(partitionId);
+            if (partition.isFinished()) {
+                return Optional.empty();
+            }
+            Optional<OpenTaskDescriptor> openTaskDescriptor = partition.getOpenTaskDescriptor();
+            if (openTaskDescriptor.isPresent()) {
+                return openTaskDescriptor.map(OpenTaskDescriptor::getNodeRequirements);
+            }
+            Optional<TaskDescriptor> taskDescriptor = taskDescriptorStorage.get(getStageId(), partitionId);
+            if (taskDescriptor.isPresent()) {
+                return taskDescriptor.map(TaskDescriptor::getNodeRequirements);
+            }
+            return Optional.empty();
+        }
+
+        public MemoryRequirements getMemoryRequirements(int partitionId)
+        {
+            StagePartition partition = getStagePartition(partitionId);
+            return partition.getMemoryRequirements();
+        }
+
+        public OutputDataSizeEstimate getOutputDataSize()
+        {
+            // TODO enable speculative execution
+            checkState(stage.getState() == StageState.FINISHED, "stage %s is expected to be in FINISHED state, got %s", stage.getStageId(), stage.getState());
+            return new OutputDataSizeEstimate(ImmutableLongArray.copyOf(outputDataSize));
+        }
+
+        public ExchangeSourceOutputSelector getSinkOutputSelector()
+        {
+            if (finalSinkOutputSelector != null) {
+                return finalSinkOutputSelector;
+            }
+            return sinkOutputSelectorBuilder.build();
+        }
+
+        public void setSourceOutputSelector(PlanFragmentId sourceFragmentId, ExchangeSourceOutputSelector selector)
+        {
+            sourceOutputSelectors.put(sourceFragmentId, selector);
+            RemoteSourceNode remoteSourceNode = remoteSources.get(sourceFragmentId);
+            verify(remoteSourceNode != null, "remoteSourceNode is null for fragment: %s", sourceFragmentId);
+            ExchangeSourceOutputSelector mergedSelector = selector;
+            for (PlanFragmentId fragmentId : remoteSourceNode.getSourceFragmentIds()) {
```

Comment:
why do we need to merge-in other fragments?

Reply:
To provide a complete selectors for all fragments read by a single remote source (keep in mind that a single remote source can read output of more than a single fragment, e.g.: UNION use case)